### PR TITLE
[WIP] Add Regular Expression challenges

### LINKED ISSUE #5794: New Curriculum:  Regular Expressions + 2 Checkpoints
We're pretty anemic in our coverage of regular expressions and we need to add more waypoints.

We should have a whole new section that is just Regex.  Target 15-20 Waypoints + 2 Checkpoints

This issue to discuss and track development of new Waypoints.

Some Brainstorms from Curriculum Development:
* Wildcard (`.`) - 1 WP
* Quantity Specifiers (`+`, `*`, and `{x,y}`) - 3 WP
   * Optionality (non Greedy Capture `?`) - 1 WP
* Character Classes `[abc]` and their Inverse - 2-3 WP
   * Character Ranges `[a-z]` and `[0-9]` - 2 WP
* Case Insensitivity Flag (`i`)
* Global Flag (`g`)
* Pre-defined character classes and their inverse (`\d`/`\D`/`w`/`\W`/`\s`/`\S`)  - 3 WP?
* Checking for Patterns (back reference) - 1 WP
* Capture Groups - 2 WP (Replace?)
* Anchors (`^` and `$`) - 2 WP

For Reference:
http://www.regular-expressions.info/tutorial.html

Waypoint Definitions:
- [ ] Using `.test`
- [ ] Literal Matches 1 - `abc`
- [ ] Literal Matches 2 - `ab|cd`
- [ ] Flags - `i` - Case Insensitive
- [ ] Flags - `g`- Global
- [ ] Wildcard - `.`
- [ ] Character Class 1 - `[abc]` - Count Vowels?
- [ ] Character Class 2 - `[abc]` ??
- [ ] Character Class 3 - `[a-z0-9]`
- [ ] Character Class 4 - `[^abc]`
- [ ] Quantity Specifiers 1 - `+`
- [ ] Quantity Specifiers 2 - `*`
- [ ] Checkpoint 1 - Match simple literal patterns with optional followers (character class)
- [ ] Anchors 1 - `^`
- [ ] Anchors 2 - `$`
- [ ] Pre-Defined Character Class 1 - `\w`
- [ ] Pre-Defined Character Class 2 - `\W`
- [ ] Pre-Defined Character Class 1 - `\d`
- [ ] Pre-Defined Character Class 1 - `\D`
- [ ] Pre-Defined Character Class 1 - `\s`
- [ ] Pre-Defined Character Class 1 - `\S`
- [ ] Quantity Specifiers 3 - `{x,y}`
- [ ] Quantity Specifiers 4 - `{x,}`
- [ ] Quantity Specifiers 5 - `{,y}`
- [ ] Optionality - `?`
- [ ] Capture Groups - `( )`
- [ ] Capture groups 2 - Using Replace?
- [ ] Checkpoint 2 - ??

## How to Contribute
* Comment on this Issue with:
   * Ideas about how to do the above curriculum
   * Write out sample descriptions or sample code
   * Claim a waypoint to work on
   * Discuss in the [CurriculumDevelopment Room](https://gitter.im/FreeCodeCamp/CurriculumDevelopment)
* Edit or create a Waypoint
   * Write PRs against [this branch](https://github.com/FreeCodeCamp/FreeCodeCamp/tree/feature/regular-expression-curriculum)

### LINKED ISSUE #9232: Introduce RegEx before "Check for Palindrome" problem
[Check for Palindrome](https://www.freecodecamp.com/challenges/check-for-palindromes) continues to be difficult to surmount for beginners in JavaScript.

This challenge has two parts to it:
- Process the input string `str` to remove non-alphanumeric characters from it.
- Use some logic to find palindromes.

This is the third challenge in **Basic Algorithmic Challenges**, and the two that comes before this, introduce the camper to recursion and String-to-Array with `reverse()`, `split()`, and `join()`.

---

In this challenge, the camper needs to come up with a non-trivial regular expression, to remove non-alphanumeric characters from the input string `str`.

And here's the kicker - `\w` does correspond to alphanumeric characters [`0-9]` and `[a-z]` (and also `[A-Z]`, but not relevant to this challenge, because we do case insensitive matching); but it also includes underscore `_`. The test cases indicate you have to exclude underscore.

---

I recommend introducing the camper to regular expression, with 1-2 problems before that, and teach them how to form and test their RegEx. At the very least, they should be able to use something like [regex101](https://regex101.com/).

Ideally, I would want a full-fledged curriculum dedicated only to RegEx. I have seen even experienced campers fear regular expressions (I am not experienced, but I fear it too!). If this issue is accepted, I can work with others on the core-team on getting a curriculum on RegEx on the live site.

I know that there are four problems that refers to regular expression:
- [Sift through text with regular expressions](https://www.freecodecamp.com/challenges/sift-through-text-with-regular-expressions)
- [Find numbers with regular expressions](https://www.freecodecamp.com/challenges/find-numbers-with-regular-expressions)
- [Find whitespaces with regular expressions](https://www.freecodecamp.com/challenges/find-whitespace-with-regular-expressions)
- [Invert regular expression matches](https://www.freecodecamp.com/challenges/invert-regular-expression-matches-with-javascript)

But none of them makes the camper comfortable forming non-trivial complex regular expressions. And in later challenges, like US telephone numbers, regular expression makes the difference between solved and unsolved. 

Besides, these four problems are from optional part of the curriculum. In the helpful links section for the palindrome problem, we can at least point them to these.

@QuincyLarson @BerkeleyTrue  @SaintPeter @CodeNonprofit FYI.

### PR DESCRIPTION
**Pull request by erictleung**
#### Pre-Submission Checklist

- [x] Your pull request targets the `staging` branch of FreeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](https://github.com/FreeCodeCamp/FreeCodeCamp/wiki/git-rebase#squashing-multiple-commits-into-one) them into one commit).
- [x] All new and existing tests pass the command `npm run test-challenges`. Use `git commit --amend` to amend any fixes.
#### Type of Change

- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)
#### Checklist:

- [x] Tested changes locally.
- [x] Closes currently open issue (replace XXXX with an issue no): Closes #5794 and Closes #9232
#### Description

Add new challenges on regular expressions based on @SaintPeter's curriculum brainstorm in #5794 and address issue brought up by @alayek in #9232.

**Proposed Challenges and Completion Status:**
- [x] Using `.test`
- [x] Match Literal String (Literal Matches 1 - `abc`)
- [x] Match Literal String with Different Possibilities (Literal Matches 2 - `ab|cd`)
- [x] Ignore Case While Matching (Flags - `i` - Case Insensitive)
- [x] Extract Matches (Introduce `.match()` method)
- [x] Find More Than First Match (Flags - `g`\- Global)
- [x] Wildcard Period (Wildcard - `.`)
- [ ] Match Single Character with Multiple Possibilities (Character Class 1 - `[abc]` - Count Vowels?)
- [ ] Match Letters of the Alphabet (Character Class 2 - `[abc]` ??)
- [ ] Match Numbers and the Alphabet (Character Class 3 - `[a-z0-9]`)
- [ ] Match Single Characters Not Specified (Character Class 4 - `[^abc]`)
- [ ] Occurs One or More Times (Quantity Specifiers 1 - `+`)
- [ ] Occurs Zero or More Times (Quantity Specifiers 2 - `*`)
- [ ] Don't Be Greedy, Be Lazy (Lazy vs greedy `.*` and `.+`)
- [ ] Checkpoint 1 **Needs new name** (Checkpoint 1 - Match simple literal patterns with optional followers (character class))
- [ ] Match Strings Only at the Beginning (Anchors 1 - `^`)
- [ ] Match Strings Only at the End (Anchors 2 - `$`)
- [ ] Use Character Classes to Match Alphanumerics (Pre-Defined Character Class 1 - `\w`)
- [ ] Use Character Classes to Match Non-Alphanumerics (Pre-Defined Character Class 2 - `\W`)
- [ ] Use Character Classes to Match Digits (Pre-Defined Character Class 1 - `\d`)
- [ ] Use Character Classes to Match Non-Digits (Pre-Defined Character Class 1 - `\D`)
- [ ] Use Character Classes to Match Whitespace (Pre-Defined Character Class 1 - `\s`)
- [ ] Use Character Classes to Match Non-Whitespace (Pre-Defined Character Class 1 - `\S`)
- [ ] Specify Upper and Lower Number of Matches (Quantity Specifiers 3 - `{x,y}`)
- [ ] Specify Only Lower Number of Matches (Quantity Specifiers 4 - `{x,}`)
- [ ] Specify Only Upper Number of Matches (Quantity Specifiers 5 - `{,y}`)
- [ ] Check For All or None (Optionality - `?`)
- [ ] Capture Groups of Patterns (Capture Groups - `( )`)
- [ ] Capture Groups and Replace (Capture groups 2 - Using Replace?)
- [ ] Checkpoint 2 **Needs new name** (Checkpoint 2 - ??)
## Do you want to help?

Choose a challenge above and please reply to this GitHub pull request with:
1. Challenge description text
2. Test suite (using the assert method)
3. The seed code, which is pre-populated in the editor at the beginning of the challenge
4. A working solution that makes all tests pass

Additionally, if you find any of the above concepts lacking in the scope of using regular expressions, please comment so that we may consider adding it to the section.

You can preview the changes in the [Files changed tab](https://github.com/FreeCodeCamp/FreeCodeCamp/pull/9286/files) above. You can add comments to pre-existing changes if you find any errors or something that could be improved.
## You want to test locally?

If you want to test these changes locally, you'll first want to have a local installation of FreeCodeCamp on your computer. Then you'll want to run these commands.

``` sh
cd path/to/your/freecodecamp # Change this according to your setup
git checkout -b erictleung-feature/add-regex-section staging
git pull https://github.com/erictleung/FreeCodeCamp.git feature/add-regex-section
```

For advanced contributors, [here](http://stackoverflow.com/a/5884825) is how to maintain a local copy of the changes to my changes.

### CODE REVIEW

File: seed/challenges/01-front-end-development-certification/regular-expressions.json [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+        "If we wanted to find the word <code>the</code> in the string <code>The dog chased the cat</code>, we could use the following regular expression: <code>/the/</code>",
+        "JavaScript has multiple ways to use regexes. One way to test a <code>regex</code> is using the <code>.test()</code> method. The <code>.test()</code> method tests the regex and will return <code>true</code> or <code>false</code> if your pattern finds something or not.",
+        "<blockquote>var testStr = \"FreeCodeCamp\";<br>var testRegex = /Code/;<br>testRegex.test(testStr); // returns true</blockquote>",
+        "<h4>Instructions</h4>",
+        "Apply the regex <code>myRegex</code> on the string <code>myString</code> using the <code>.test()<code> method."
+      ],
+      "challengeSeed": [
+        "var myString = \"Hello, World!\";",
+        "var myRegex = /Hello/;",
+        "myRegex; // Change this line"
```
Comment (SaintPeter) [inline seed/challenges/01-front-end-development-certification/regular-expressions.json]: We might want to assign this to a result so we can test it:
```
var result = myRegex.test(myString);
```

**Comment by SaintPeter** [comment]
It might make sense to create sort of a basic testing framework that applies a regex using `.test` with a standard input and a standard output.

Something like:
```js
var testTarget = "Whatever we're going to test here";
var myRegExp = /test/; // Change This line
var result = myRegExp.test(testTarget);
```

or maybe wrap it in a function?

```js
function regexTester(testString) {
  var myRegExp = /test/; // Change this line only
  return myRegExp.test(testString);
}

// Change below for testing
regexTester("This is a test");
```
Then we can run multiple tests against the function with various input.

We should probably introduce `.match` as well, so we can count the output and/or do array comparisons.

**Comment by erictleung** [comment]
@SaintPeter the section is focused on regexes so I think we could just take the regex and apply is to the string or strings we want. So assert statements for testing could be something like this:

```js
assert(myRegExp.test("Test this string"), 'message: ....');
```

But if people want to use functions, I'm fine by that too. 

I agree that we should also introduce `.match` as well. I'll have to think about where to put it. We could either put it near the beginning, or near the end.

Side note, some regular expression links that might prove useful in developing this section.

- Regex crossword https://regexcrossword.com/
- Regex golf http://regex.alf.nu/
- The Case for Regular Expressions https://technet.microsoft.com/en-us/magazine/ee835717.aspx
- Regular Expressions http://www.regular-expressions.info/
- The Many Uses of Regex http://www.rexegg.com/regex-uses.html
- Regular Expression Examples https://en.wikipedia.org/wiki/Regular_expression#Examples

**Comment by QuincyLarson** [comment]
@erictleung Should we also show them how to use lookarounds ([ahead and behind](http://www.regular-expressions.info/lookaround.html)) and [laziness/greediness](http://www.regular-expressions.info/repeat.html)? I have to use those occasionally, and if we're going to ever cover them, this is probably the place.

**Comment by erictleung** [comment]
I'm putting in code blocks and want to indicate what a certain line of code will output. I found putting the output on a new line is nicer so you don't have to scroll (generally speaking for devices).

**New way:**
image

**Old way:**
image

**Another new way:**
image

Thoughts?

**Comment by alayek** [comment]
I prefer the "New Way", over "Another New Way". The new way clearly states that it returns an array of strings.

**Comment by atjonathan** [comment]
I agree with @alayek :+1:, another new way isn't clear enough.

**Comment by erictleung** [comment]
Closing until infrastructure is ready. Moving discussion over to https://github.com/FreeCodeCamp/CurriculumExpansion/issues/22.

**Pull request closed without merging**